tidy3d/web/api/autograd/autograd.py

```
@@ -988,78 +988,78 @@ def postprocess_adj(
-        struct_bounds = rmin_struct, rmax_struct = structure.geometry.bounds
-        rmin_sim, rmax_sim = sim_data_orig.simulation.bounds
-        rmin_intersect = tuple([max(a, b) for a, b in zip(rmin_sim, rmin_struct)])
-        rmax_intersect = tuple([min(a, b) for a, b in zip(rmax_sim, rmax_struct)])
-        bounds_intersect = (rmin_intersect, rmax_intersect)
-
-        derivative_info = DerivativeInfo(
-            paths=structure_paths,
-            E_der_map=E_der_map.field_components,
-            D_der_map=D_der_map.field_components,
-            E_fwd=E_fwd.field_components,
-            E_adj=E_adj.field_components,
-            D_fwd=D_fwd.field_components,
-            D_adj=D_adj.field_components,
-            eps_data=eps_fwd.field_components,
-            eps_in=eps_in,
-            eps_out=eps_out,
-            eps_background=eps_background,
-            frequency=freq_adj,
-            eps_no_structure=eps_no_structure,
-            eps_inf_structure=eps_inf_structure,
-            bounds=struct_bounds,
-            bounds_intersect=bounds_intersect,
-        )
+            vjp_value_map = structure.compute_derivatives(derivative_info)
 
-        vjp_value_map = structure.compute_derivatives(derivative_info)
-
-        # extract VJPs and put back into sim_fields_vjp AutogradFieldMap
-        for structure_path, vjp_value in vjp_value_map.items():
-            sim_path = tuple(["structures", structure_index] + list(structure_path))
-            sim_fields_vjp[sim_path] = vjp_value
+            # extract VJPs and put back into sim_fields_vjp AutogradFieldMap
+            for structure_path, vjp_value in vjp_value_map.items():
+                sim_path = tuple(["structures", structure_index] + list(structure_path))
+                if freq_idx == 0:
+                    sim_fields_vjp[sim_path] = vjp_value
+                else:
+                    sim_fields_vjp[sim_path] += vjp_value
```

Not necessarily related to this PR but I wonder if `vjp_value` can ever be something other than a scalar or a numpy array, in which case this would silently fail. We ran into this in #2319. Might be good to double check vs `Box` center & size derivatives because that method does return tuples for the VJPs, see [here](https://github.com/flexcompute/tidy3d/blob/74d5be32de7a15670273416c50246608cf6cc45c/tidy3d/components/geometry/base.py#L2520-L2523)...

---

Good point, didn't think of this when implementing. Tried with the box to see how this would ultimately happen in the code and I think this only operates on scalar values in that case. The center/size derivatives come back as a dictionary of tuples as you pointed out, but those get packaged up into a dictionary that identifies each gradient by its component in either center or size. That dictionary has keys that look like ('geometry', 'size', 0) which select out the attribute and index into the attribute for the derivative and then the vjp value is just a scalar.

For the vertices, this ends up looking like ('geometry', 'vertices') and then the vertices are being summed. If those vertices came back as a list or tuple (or some object using this code path had that) though instead of an array, this could be a problem. I've added a change to catch this situation. One thing I'm wondering about the solution from the #2319 that didn't occur to me before is if the list is not flat. For example,

>>> ` x = [ [ 1, 2 ], [ 3, 4 ] ]`
>>> `y = [ [ 5, 6 ], [ 7, 8 ] ]`
>>> `type(x)(x + y for x, y in zip(x, y))`
[[1, 2, 5, 6], [3, 4, 7, 8]]

instead of [[6, 8], [10, 12]]

---

yeah you're right. I guess the question is, do we ever run into nested tuples? if all tuples map to regular arrays, we could cast to numpy, to the math there, and then cast back. otherwise we'd have to recursively map through the nesting. I feel like that would be over-engineering it if there are never nested tuples in the vjps in the first place

---

Yeah, not sure if we have a lot of possibility for it. The reason it came up in my head is the vertices are in a nested form when they come into this function (they are a list of coordinate pairs). They are not coming in as a list/tuple and are summing completely fine as arrays, so not sure if there's a case where things could possibly come in that way. I would assume that at that point, there would be a dimension mismatch down the line and some sort of error.